### PR TITLE
Activity Log: Rewind

### PR DESCRIPTION
**Pull request by elibud**
**Another step for** #7832

Done on top of #8225 and not to be merged until that one is.

**To test:**

1. Go to the activity log and find a rewindable activity (blue ribbon on the left).
2. Tap on it and select Rewind.
3. Wait for it.
4. Wait for it.
5. Wait for it.
6. Check on the web that the site was properly rewound.
7. Leave the screen and come back and check on the phone that the site was probably rewound.

Notes:
As with this series of PRs this is the simplest UI to make the feature work, not intended for actual use by human beings.

### CODE REVIEW

**Comment by koke** [comment]
One thing I noticed is that the rewind process isn't blocking, so you can request a rewind, and while you're waiting scroll and request a second one 😬 

simulator screen shot - iphone x - 2017-11-28 at 17 28 15

**Comment by koke** [comment]
Besides being able to trigger multiple rewinds manually, my concern with the polling implementation is that it doesn't have any checks to prevent a request if one's in progress, so if a `restoreStatusForSite` takes more than 5 seconds, there will be two going at the same time.

An alternative would be to not have a repeating timer, and only schedule the next one when the `restoreStatusForSite` call finishes (unless it's completed).

I implemented a related thing for `PingHubManager` with `IncrementalDelay` and `DispatchDelayedAction`.

Also, I think this restore state logic is the perfect candidate to create an `ActivityStore`/`ActivityService` 😉

**Comment by elibud** [comment]
>  trigger multiple rewinds manually, 
Fixed! Thanks for catching that one, I think I was on a very optimistic day when I did this one :)

> concerns with polling implementation
I've followed your approach on the PingHubManager, really liked it, I should've guessed we had something already done to sort this things out.

> Also, I think this restore state logic is the perfect candidate to create an ActivityStore/ActivityService 😉
For sure! It's just further down the list. On the last set of PRs I'm just trying to test that we have all the endpoints+info we need and that they work as expected. There's also some changes to the API that are coming soon, so I'm waiting for them to migrate all this to Flux.

There's also the problem of the UI refresh after the rewind is done, currently when I fetch the activities after a rewind has completed, the successful rewind takes a few seconds to get to the feed, so the rewind is not represented correctly, but I'm expecting this to be fixed server side (fingers crossed).

So, in summary, ready for another look @koke, but keep in mind this is still very much a WIP.

File: WordPress/Classes/ViewRelated/Activity/ActivityListViewController.swift [hunk 1]
```diff
@@ -104,3 +126,103 @@ extension ActivityListViewController: WPNoResultsViewDelegate {
...
+
+extension ActivityListViewController {
+
+    struct RestoreStatusCheck {
+        static let maxRetries = 12
+        static let pollingInterval = 5.0
+    }
+
+    fileprivate func restoreSiteToRewindID(_ rewindID: String) {
+        tableView.isUserInteractionEnabled = false
```
Comment (koke) [inline WordPress/Classes/ViewRelated/Activity/ActivityListViewController.swift]: I was going to suggest that SVProgressHUD already has blocking built in via `SVProgressHUD.setDefaultMaskType(.black)`, but that blocks the whole screen and the user can't event navigate away until the process is complete. I'm not sure if that's what we want, so your approach is probably better.

simulator screen shot - iphone x - 2017-12-04 at 14 48 01

Comment (elibud) [reply WordPress/Classes/ViewRelated/Activity/ActivityListViewController.swift]: I'm sure once design is done with this one that hud will be gone to be honest.

File: WordPress/Classes/ViewRelated/Activity/ActivityListViewController.swift [hunk 2]
```diff
@@ -1,20 +1,29 @@
...
 
-    @objc let siteID: Int
-    @objc let service: ActivityServiceRemote
+    let siteID: Int
+    let service: ActivityServiceRemote
+
+    enum Configuration {
+        /// Sequence of increasing delays to apply to the fetch restore status mechanism (in seconds)
+        ///
+        static let delaySequence = [1, 5, 10, 15, 20, 25, 30]
```
Comment (koke) [inline WordPress/Classes/ViewRelated/Activity/ActivityListViewController.swift]: I think this might be too much waiting? Also, I don't think `IncrementalDelay` was my best effort at naming things, but this delay will repeat forever, so `showErrorFetchingRestoreStatus` never happens now.

I think `DispatchDelayedAction` works better than the timer here, but your initial approach of a single 5 second delay up to 12 times was best.

Comment (elibud) [reply WordPress/Classes/ViewRelated/Activity/ActivityListViewController.swift]: I'm going to add a parameter to the `IncrementalDelay` constructor to allows us to make it repeating or not and this should look better.

Comment (koke) [reply WordPress/Classes/ViewRelated/Activity/ActivityListViewController.swift]: The problem is that `IncrementalDelay` is designed for an infinite sequence (that's why I say I named it wrong), and will assume that there's always a `next()` item.

Comment (elibud) [reply WordPress/Classes/ViewRelated/Activity/ActivityListViewController.swift]: yeah I got that, I'm looking into how hard would it be to modify it to support both infinite and finite sequences.

**Comment by elibud** [comment]
@koke can you take another quick look? Thanks

File: WordPress/Classes/ViewRelated/Activity/ActivityListViewController.swift [hunk 3]
```diff
@@ -178,6 +175,12 @@ extension ActivityListViewController {
     }
 
     fileprivate func checkStatusDelayedForRestoreID(_ restoreID: String) {
+        delayedRetryAttempt = delayedRetryAttempt + 1
+        guard delayedRetryAttempt < Configuration.maxRetries else {
+            delay.reset()
+            showErrorFetchingRestoreStatus()
```
Comment (koke) [inline WordPress/Classes/ViewRelated/Activity/ActivityListViewController.swift]: Should this also set `tableView.isUserInteractionEnabled = true`?

Comment (elibud) [reply WordPress/Classes/ViewRelated/Activity/ActivityListViewController.swift]: Yup, added and merging. Thanks for the review :)
